File: lib/services/content_regeneration_service.dart
import 'package:flutter/foundation.dart';

import '../core/app_language.dart';
import '../data/db/drift_database.dart';
import 'meal_template_generator.dart';
import 'profile_fit.dart';
import 'user_profile_service.dart';
import 'workout_template_generator.dart';

/// What a regeneration would replace, so the prompt can be specific about
/// the cost before the user agrees to it.
class RegenerationPreview {
  const RegenerationPreview({
    required this.mealTemplates,
    required this.workoutTemplates,
  });

  final int mealTemplates;
  final int workoutTemplates;

  bool get isEmpty => mealTemplates == 0 && workoutTemplates == 0;
}

/// Replaces profile-generated templates after the profile changes.
///
/// The rule this exists to enforce: **only `TemplateOrigin.generated` is
/// ever touched.** Anything the user built or edited is theirs, and the
/// seeded built-ins are left alone so the catalog can't shrink
/// irreversibly. Without that distinction, "your equipment changed --
/// regenerate?" has no safe implementation; it would either do nothing or
/// destroy hand-built work.
///
/// Deliberately *not* automatic. Regeneration throws away templates the
/// user may have been using all week, so it is offered, not imposed --
/// [ProfileFit.contentAffectingFieldsChanged] decides whether to offer it
/// at all, and only content-affecting fields count (`goal` is target-only
/// and does not qualify).
class ContentRegenerationService {
  ContentRegenerationService(this._database);

  final AppDatabase _database;

  /// How many generated templates a regeneration would discard.
  Future<RegenerationPreview> preview() async {
    final meals = (await _database.getAllMealTemplates())
        .where((t) => ProfileFit.isReplaceable(t.origin))
        .length;
    final workouts = (await _database.getAllWorkoutTemplates())
        .where((t) => ProfileFit.isReplaceable(t.origin))
        .length;
    return RegenerationPreview(
      mealTemplates: meals,
      workoutTemplates: workouts,
    );
  }

  /// Deletes every generated template and rebuilds from [profile], in
  /// [language].
  ///
  /// [language] is the app's *current* language, not the one the discarded
  /// templates were written in. Regeneration is the one moment a rewrite is
  /// legitimate: the user asked for these templates to be replaced, so the
  /// replacements are written in the language they are using now. Nothing
  /// else in the app re-languages content behind their back.
  ///
  /// Returns how many templates were created. Safe to call when nothing was
  /// previously generated -- it simply generates for the first time.
  Future<int> regenerate(UserProfile profile, AppLanguage language) async {
    final removedMeals = await _deleteReplaceableMealTemplates();
    final removedWorkouts = await _deleteReplaceableWorkoutTemplates();
    debugPrint('[REGEN] Removed $removedMeals meal / $removedWorkouts workout '
        'generated templates');

    final meals = await MealTemplateGenerator(_database, profile, language)
        .generateTemplates();
    final workouts =
        await WorkoutTemplateGenerator(_database, profile, language)
            .generateTemplates();

    debugPrint('[REGEN] Rebuilt ${meals.length} meal / ${workouts.length} '
        'workout templates');
    return meals.length + workouts.length;
  }

  Future<int> _deleteReplaceableMealTemplates() async {
    final replaceable = (await _database.getAllMealTemplates())
        .where((t) => ProfileFit.isReplaceable(t.origin))
        .toList();
    for (final template in replaceable) {
      // deleteMealTemplate cascades to its items.
      await _database.deleteMealTemplate(template.id);
    }
    return replaceable.length;
  }

  Future<int> _deleteReplaceableWorkoutTemplates() async {
    final replaceable = (await _database.getAllWorkoutTemplates())
        .where((t) => ProfileFit.isReplaceable(t.origin))
        .toList();
    for (final template in replaceable) {
      await _database.deleteWorkoutTemplate(template.id);
    }
    return replaceable.length;
  }
}

File: Packages/WellnessKit/Sources/WellnessServices/ContentRegenerationService.swift
import Foundation
import WellnessModels
import WellnessCatalog
import WellnessDomain
import WellnessPersistence

public struct RegenerationPreview: Sendable {
    public let mealTemplates: Int
    public let workoutTemplates: Int
    public var isEmpty: Bool { mealTemplates == 0 && workoutTemplates == 0 }
}

public struct ContentRegenerationService: Sendable {
    private let food: any FoodStore
    private let mealTemplate: any MealTemplateStore
    private let workoutTemplate: any WorkoutTemplateStore
    private let exercise: any ExerciseStore

    public init(
        food: any FoodStore,
        mealTemplate: any MealTemplateStore,
        workoutTemplate: any WorkoutTemplateStore,
        exercise: any ExerciseStore
    ) {
        self.food = food
        self.mealTemplate = mealTemplate
        self.workoutTemplate = workoutTemplate
        self.exercise = exercise
    }

    public func preview() async throws -> RegenerationPreview {
        let meals = try await mealTemplate.allMealTemplates()
            .filter { ProfileFit.isReplaceable($0.origin) }.count
        let workouts = try await workoutTemplate.allWorkoutTemplates()
            .filter { ProfileFit.isReplaceable($0.origin) }.count
        return RegenerationPreview(mealTemplates: meals, workoutTemplates: workouts)
    }

    @discardableResult
    public func regenerate(profile: UserProfile, language: AppLanguage) async throws -> Int {
        try await deleteReplaceableMealTemplates()
        try await deleteReplaceableWorkoutTemplates()

        let allFoods = try await food.allFoods()
        let allExercises = try await exercise.allExercises()

        let catalogNameById = Dictionary(
            uniqueKeysWithValues: Catalog.foods.map { ($0.id, $0.name) }
        )

        var counter = 0

        let mealGen = MealTemplateGenerator(profile: profile, language: language)
        let mealResults = mealGen.generateTemplates(
            availableFoods: allFoods,
            catalogNameById: catalogNameById,
            makeId: { UUID().uuidString }
        )
        for result in mealResults {
            var t = result.template
            t.items = result.items
            try await mealTemplate.insertMealTemplate(t)
            for item in result.items {
                try await mealTemplate.insertMealTemplateItem(item)
            }
            counter += 1
        }

        let workoutGen = WorkoutTemplateGenerator(profile: profile, language: language)
        let workoutResults = workoutGen.generateTemplates(
            allExercises: allExercises,
            makeId: { UUID().uuidString }
        )
        for (template, exercises) in workoutResults {
            var t = template
            t.exercises = exercises
            try await workoutTemplate.insertWorkoutTemplate(t)
            for ex in exercises {
                try await workoutTemplate.insertTemplateExercise(ex)
            }
            counter += 1
        }

        return counter
    }

    private func deleteReplaceableMealTemplates() async throws {
        let all = try await mealTemplate.allMealTemplates()
        for t in all where ProfileFit.isReplaceable(t.origin) {
            try await mealTemplate.deleteMealTemplate(id: t.id)
        }
    }

    private func deleteReplaceableWorkoutTemplates() async throws {
        let all = try await workoutTemplate.allWorkoutTemplates()
        for t in all where ProfileFit.isReplaceable(t.origin) {
            try await workoutTemplate.deleteWorkoutTemplate(id: t.id)
        }
    }
}
